### ISSUE TITLE
TSLint allow-leading-underscore

### ISSUE THREAD

**Issue by vilimco** [degordian/js-coding-standards#1]
Current config does not have this flag turned on.

This is an example of how getters/setters are used in TS/JS:
```js
class A {
    private _name : string;

    constructor(name : string) {
        this._name = name;   
    }

    get name() : string {
        return this._name;
    }
}
```

For someone coming from strictly typed languages, this is weird, since approach in those languages would be like this:
```js
class A {
    private name : string;

    constructor(name : string) {
        this.name = name;   
    }

    getName() : string {
        return this.name;
    }
}
```

Problem is that default generator for getters/setter in IDEs will generate them as in the first example since this is native behaviour and then TSLint marks this as an error.

Since we are going to use this attribute normally throughout the code (without leading underscore) and we are using generators for getters/setters, this should be turned on.

We could switch to standard getAttr() format, but that is forcing other languages syntax to JS. 
Then we should somehow disable get/set in this native form (through custom rules or something similar) to be consistent all over the projects. That, somehow, feels like an overkill and too much forcing of Java/C# syntax here.

Also, situations like this will happen oftenly without the usage of IDEs generators:
```js
class A {
    private name : string;

    constructor(name : string) {
        this.name = name;   
    }

    get name() : string {
        return this.name;
    }
}
```
This will result in stack overflow due to its recursive callback to itself (generators change access to the attribute and rename it properly so you don't get this situation).

**Comment by krukru** [comment]
@degordian/team-developers please discuss

**Comment by leovujanic** [comment]
Why not have both? What is the blocker? From my point of view, whatever you can do in some language is "language way" JS way is just JS way and you can do whatever you want. :D

Why I think we need to have both is because of the following case. We want to print name with some prefix - we can't do that with `get`
```javascript
class A {
    
    private _name: string;

    private _title: string;

    get name(withTitle: bool): string {
        if (withTitle) {
            return this._title + this._name;
        }

        return this._name;
    }

    set name(value) {
        this._name = value;
    }
}
```
giving error `Getter must not have any formal parameters`.

On the other hand, get and set are not guaranty that you will receive right type. 
```javascript
set name(value) {
        this._name = value;
}
```

So for me its a something that will depend on the use case, and of course there is responsibility on developer not to have property and get with the same name.

**Comment by bonzzy** [comment]
@leovujanic I think it is not natural to have any input params in getters. 
In C# getters and setters use private parameters with underscore just like in Typescript. 
Example of class in C#:
```csharp
class Example
{
    bool _found;
    public bool IsFound
    {
        get
        {
            return this._found;
        }
        private set
        {
            // Can only be called in this class.
            this._found = value;
        }
    }
}
```
For example Dart doesn’t have the keywords public, protected, and private. If an identifier starts with an underscore `_`, it’s private to its library.
I would recommend to follow language principles and use underscore with private parameter if using getters and setters.
There was a similar discuss here [https://github.com/angular/angular.io/issues/1108](https://github.com/angular/angular.io/issues/1108).

**Comment by leovujanic** [comment]
@bonzzy Of, course. What I was trying to say is that we should adjust config and use generators for `get` and `set` (with underscores). We do not need use one against the other or say we will use `getAttr()` instead of get. We should use both, depending on your needs. Property names with underscores are the convention, not what we will use for getters.

Take a look at this example.  It's working and is the legit thing to do.
```javascript
class A {

    private name: string;

    private lname: string;

    get fullName(): string {
        return this.name + this.lname;
    }
}

const a = new A();
console.log(a.fullName);
```
I think the main question here is "Should we prefix all private properties with underscore?".

**Comment by vilimco** [comment]
@leovujanic 
As it regards getter with boolean, it is actually best practice to separate those two into two different methods (regardless if it is getter or not, especially for simple functions like that). It moves the responsibility from the class to the caller :)

As it goes for naming: no, we should definitely not name all private variables with _. That is not the point here.

Point is that if you are using native getter/setter for attribute, it should be prefixed with _ to avoid recursive callbacks. As I said earlier, alternative to this is to have all getter/setter functions named as getXXX/setXXX as it is done in other languages.

Using get/set for "non existing" attributes that are composed from other attributes is completely ok and I don't see any problem with that.

You can always write getXXX, that is completely ok too. 

But for the sake of native getters/setters and how they are being used most of the time in JS, we need this flag turned on in TSLint.

**Comment by krukru** [comment]
C# recommends avoid using underscore to prefix private members, and to use PascalCase for naming getters. Obviously this would not fit here, we do not use PascalCase other than for ClassName. Most often C# doesn't even need private member variables using this syntax sugar 
```C#
class Example
{
    public bool IsFound { get; private set; }
}
```

Strictly speaking getters (using `get` construct) should only be used for retuning value with minimal logic. Validation or stuff like string concatenation is fine. Things that actually do some work should be called `getDatabaseConnection()` without the `get` construct. Somewhere I found that this is a unwritten convention. By using `get` you are telling that this is a cheap operation. I cannot find this article, but I like the idea.

**Comment by krukru** [comment]
Which brings us to the problem at hand. We have a contradiction. If we do not prefix with `_`, we cannot have this snippet of code 
```typescript
class A {
    private name : string;

    constructor(name : string) {
        this.name = name;   
    }

    get name() : string {
        return this.name;
    }
}
```
The getter should be named `get Name` which I do not like since it sticks out too much (we never use PascalCase). Or we can prefix with `_`, but this is against Palantir's tslint recommended set of rules. I will check how they solve this collision, for now I am leaning towards overriding their rule and force `_` for private members

**Comment by krukru** [comment]
Solved with #5. We now use underscore for all private members
